signature/src/encoding.rs

```diff
@@ -0,0 +1,38 @@
+//! Encoding support.
+
+use crate::{Error, Result};
+
+#[cfg(feature = "alloc")]
+use alloc::{boxed::Box, vec::Vec};
+
+/// Support for decoding/encoding signatures as bytes.
+pub trait SignatureEncoding:
+    Clone + Sized + for<'a> TryFrom<&'a [u8], Error = Error> + Into<Self::Repr>
+{
+    /// Byte representation of a signature.
+    type Repr: 'static + AsRef<[u8]> + AsMut<[u8]> + Clone + Default + Send + Sync;
+
+    /// Decode signature from its byte representation.
+    fn from_bytes(bytes: &Self::Repr) -> Result<Self> {
+        Self::try_from(bytes.as_ref())
+    }
+
+    /// Encode signature as its byte representation.
+    fn to_bytes(&self) -> Self::Repr {
+        self.clone().into()
+    }
+
+    /// Encode signature as a byte vector.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    fn to_vec(&self) -> Vec<u8> {
+        self.to_bytes().as_ref().to_vec()
+    }
+
+    /// Encode the signature as a boxed byte slice.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    fn to_boxed_slice(&self) -> Box<[u8]> {
```

Comment:
In https://github.com/RustCrypto/signatures/pull/562/files#diff-23f3fe37b15706957a041307578d2433204986a9eb42dc8d7c4126dcaf9266f1R226-R239 I opted for `Box<[u8]>` as the `Repr` for most of the ASN.1 or otherwise variable-width types. In addition to `ecdsa::der::Signature` there's also `dsa::Signature`:

https://github.com/RustCrypto/signatures/pull/562/files#diff-9d900cc36738a78e813145f657efa63ba433146f1046ff9104d4a943b009ac87R95

The main advantage of a boxed slice over a `Vec` is that it eliminates the capacity member of the underlying struct, and thus makes such types slightly smaller.

Reply:
That said, it might make sense to have fewer methods, and names like `to_vec` and `to_boxed_slice` can clash with other traits easily, requiring explicit disambiguation using e.g. `<Self as SignatureEncoding>::` or `(self as SignatureEncoding)` which is annoying.

Reply:
Agreed. Not very familiar with embedded systems, but it seems that adding a method to a 3-method trait is pretty heavy just to save 1 usize.

Reply:
Yeah, I can try removing it to simplify the trait